File: DesignPatterns/01_GOF23/DP_examples/C/01_Creational/Factory/product.c
#include <stdio.h>
#include <stdlib.h>
#include "product.h"

// 圆形产品方法
static void circleDraw(IShape *self)
{
    printf("Draw a Circle\n");
}

static void circleFill(IShape *self)
{
    printf("Fill red\n");
}

// 正方形产品方法
static void squareDraw(IShape *self)
{
    // use self->data
    printf("Draw a Square\n");
}

static void squareFill(IShape *self)
{
    printf("Fill Green\n");
}

// 矩形产品方法
static void rectangleDraw(IShape *self)
{
    printf("Draw a Rectangle\n");
}

static void rectangleFill(IShape *self)
{
    printf("Fill Blue\n");
}

// 未知形状产品方法
static void unknownShapeDraw(IShape *self) {}

static void unknownShapeFill(IShape *self) {}

// shape create
void shape_destroy(IShape *self)
{
    if (self == NULL)
        return;
    if (self->data != NULL)
    {
        free(self->data);
        self->data = NULL;
        self->draw = NULL;
        self->fill = NULL;
    }
    free(self);
}
IShape *createShape(ShapeKind kind)
{
    IShape *shape = (IShape *)malloc(sizeof(IShape));
    switch (kind)
    {
    case CIRCLE:
        shape->data = malloc(sizeof(CircleData));
        shape->draw = circleDraw;
        shape->fill = circleFill;
        break;
    case SQUARE:
        shape->data = malloc(sizeof(SquareData));
        shape->draw = squareDraw;
        shape->fill = squareFill;
        break;
    case RECTANGLE:
        shape->data = malloc(sizeof(RectangleData));                
        shape->draw = rectangleDraw;
        shape->fill = rectangleFill;
        break;
    default:
        shape->data = malloc(sizeof(UnknownShapeData));
        shape->draw = unknownShapeDraw;
        shape->fill = unknownShapeFill;
        break;
    }
    shape->destroy = shape_destroy;
    return shape;
}

File: 01_GOF23/DP_examples/Lua/01_Creational/Factory/Product.lua
-- 产品类型枚举

---@enum ShapeKind
ShapeKind = {
    UNKNOWN_SHAPE = 0,
    CIRCLE = 1,
    SQUARE = 2,
    RECTANGLE = 3,
    BIG_CIRCLE = 4
}

-- 产品接口
---@class IShape
IShape = {
    draw = function(self)
        error("Subclass must implement draw method")
    end,
    fill = function(self)
        error("Subclass must implement fill method")
    end
}

-- 创建产品对象的函数
---@return IShape
function IShape:new()
    local o = {}
    setmetatable(o, {__index = self})
    return o
end

-- 圆形产品
---@class Circle : IShape
Circle = IShape:new()
function Circle:draw()
    print("Draw a Circle")
end
function Circle:fill()
    print("Fill red")
end

-- 正方形产品
---@class Square : IShape
Square = IShape:new()
function Square:draw()
    print("Draw a Square")
end
function Square:fill()
    print("Fill Green")
end

-- 矩形产品
---@class Rectangle : IShape
Rectangle = IShape:new()
function Rectangle:draw()
    print("Draw a Rectangle")
end
function Rectangle:fill()
    print("Fill Blue")
end

-- 未知形状产品
---@class UnknownShape : IShape
UnknownShape = IShape:new()
function UnknownShape:draw()
    -- 空实现
end
function UnknownShape:fill()
    -- 空实现
end
